### ISSUE TITLE
[4.0] refactor toolbar code

### ISSUE THREAD

**Issue by brianteeman** [joomla/joomla-cms#18392]
If you look at the toolbar helper and the way that we implement the icon selection for a custom toolbar item and the way we determine an id for the button you will see that it is based on an image name eg 

`\JToolbarHelper::custom('searches.reset', 'refresh.png', 'refresh_f2.png', 'JSEARCH_RESET', false);
`
IIRC this was for backwards compatibility with J2.5. As we dont have those images in core the code needs to be refactored to just use an icon and id name AND we need to be able to have an id that is not based on the icon so that we can avoid duplicate id on the page just because we have the same icon used more than once

**Comment by dgrammatiko** [comment]
I've already started refactoring the toolbar with basic aims:
- buttons context aware
- no inline javascript
- CSR (client side rendering)
- more flexible, easier customisable
- 100% B/C

Check the code: https://github.com/joomla/joomla-cms/compare/4.0-dev...dgt41:§4.0-dev-toolbar-cleanup?expand=1

NOTE: That branch is effectively frozen for some time now awaiting UX devisions

**Comment by brianteeman** [comment]
I guess that we will be waiting until hell freezes over then ;)

**Comment by dgrammatiko** [comment]
Let's give them the time they need to finish the research/tests, it's not even 3 months
![screen shot 2017-10-23 at 12 34 39](https://user-images.githubusercontent.com/3889375/31882228-bd0118b2-b7ee-11e7-98cc-6908a4098e7e.png)

**Comment by brianteeman** [comment]
Have they started?

**Comment by C-Lodder** [comment]
You do make me laugh, @brianteeman

**Comment by brianteeman** [comment]
@dgt41 are you still waiting on ux decisions for this toolbar refactor or has the refactoring restarted?

**Comment by dgrammatiko** [comment]
@brianteeman still on hold till we finalise the specs

**Comment by asika32764** [comment]
I have some idea very similar to https://github.com/joomla/joomla-cms/pull/18501 on only php part and still 100% B/C

```php
$toolbar = $container->get('toolbar');

// First button
$toolbar->addNew('JTOOLBAR_NEW')
    ->addClass('hasTooltip btn-success')
    ->setIcon('ok')
    ->setTask('article.new')
    ->setDisabled(!$canDo->core_create);

// Second button
$toolbar->addPublished('JTOOLBAR_PUBLISHED')
    -> //...

// B/C
$toolbar->addButton('Custom', ...);
```

**Comment by brianteeman** [comment]
closed as in progress elsewhere
